### ISSUE TITLE
Issue when trying to read data from memory

### ISSUE THREAD

**Issue by wshdf2010** [BVLC/caffe#2111]
I used Cifar10 netwrok, I tried to read image data using opencv and make prediction based on it.
This is the C++ code:

#include <iostream>
#include  <vector>
#include <cuda_runtime.h>
#include <cstring>
#include <cstdlib>
#include <vector>
#include <string>
#include <iostream>
#include <stdio.h>
#include "caffe/caffe.hpp"
#include "caffe/util/io.hpp"
#include "caffe/blob.hpp"
#include "caffe/data_layers.hpp"

#include "opencv2/core/core.hpp"
#include "opencv2/highgui/highgui.hpp"

using namespace caffe;
using namespace std;

int main()
{
	Caffe::set_mode(Caffe::GPU);

	string prototxt_filename="./cifar10_quick_train_test.prototxt";
	string pretrained_model="./cifar10_quick_iter_40000.caffemodel";
	Net<float> net(prototxt_filename,caffe::TEST);
}

This is the prototxt file; cifar10_quick_train_test.prototxt
name: "CIFAR10_quick"
layer {
  name: "cifar"
  type: "MemoryData"
  top: "data"
  top: "label"
  memory_data_param {
    batch_size: 1
    channels: 1
    height: 32
    width: 32
  }
  transform_param 
  {
    mean_file: "./training_image_32_mean.binaryproto"
    mirror: false
  }
}
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "data"
  top: "conv1"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 32
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.0001
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "conv1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "relu1"
  type: "ReLU"
  bottom: "pool1"
  top: "pool1"
}
layer {
  name: "conv2"
  type: "Convolution"
  bottom: "pool1"
  top: "conv2"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 64
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu2"
  type: "ReLU"
  bottom: "conv2"
  top: "conv2"
}
layer {
  name: "pool2"
  type: "Pooling"
  bottom: "conv2"
  top: "pool2"
  pooling_param {
    pool: AVE
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "conv3"
  type: "Convolution"
  bottom: "pool2"
  top: "conv3"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 128
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "relu3"
  type: "ReLU"
  bottom: "conv3"
  top: "conv3"
}
layer {
  name: "pool3"
  type: "Pooling"
  bottom: "conv3"
  top: "pool3"
  pooling_param {
    pool: AVE
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "ip1"
  type: "InnerProduct"
  bottom: "pool3"
  top: "ip1"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 64
    weight_filler {
      type: "gaussian"
      std: 0.1
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "ip2"
  type: "InnerProduct"
  bottom: "ip1"
  top: "ip2"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  inner_product_param {
    num_output: 2
    weight_filler {
      type: "gaussian"
      std: 0.1
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "accuracy"
  type: "Accuracy"
  bottom: "ip2"
  bottom: "label"
  top: "accuracy"
}
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "ip2"
  bottom: "label"
  top: "loss"
}

when I run it. I came across this problem:( the last several lines)  
I0312 11:32:16.266396  4125 net.cpp:205] This network produces output accuracy
I0312 11:32:16.266399  4125 net.cpp:205] This network produces output loss
I0312 11:32:16.266410  4125 net.cpp:447] Collecting Learning Rate and Weight Decay.
I0312 11:32:16.266415  4125 net.cpp:217] Network initialization done.
I0312 11:32:16.266417  4125 net.cpp:218] Memory required for data: 422188
W0312 11:32:16.269310  4125 net.cpp:811] Unknown layer name data
Caffe_testing: /usr/include/boost/smart_ptr/shared_ptr.hpp:653: typename boost::detail::sp_member_access<T>::type boost::shared_ptr<T>::operator->() const [with T = caffe::MemoryDataLayer<float>; typename boost::detail::sp_member_access<T>::type = caffe::MemoryDataLayer<float>*]: Assertion `px != 0' failed.

Thanks ahead for your help!
best.

**Comment by wshdf2010** [comment]
already post this issue in caffe user group. 

thanks

**Comment by n-zhang** [comment]
It seems that there is no source file  in your data layer definition so that the net doesn't know where to look for the data. Please refer to the extensive documentation, tutorials, and examples at http://caffe.berkeleyvision.org/
